kedro/config/omegaconf_config.py

```diff
@@ -0,0 +1,258 @@
+
+        # Load chosen env config
+        run_env = self.env or self.default_run_env
+        env_path = str(Path(self.conf_source) / run_env)
+        env_config = self.load_and_merge_dir_config(
+            env_path, [*self.config_patterns[key]]
+        )
+
+        # Destructively merge the two env dirs. The chosen env will override base.
+        common_keys = config.keys() & env_config.keys()
+        if common_keys:
+            sorted_keys = ", ".join(sorted(common_keys))
+            msg = (
+                "Config from path '%s' will override the following "
+                "existing top-level config keys: %s"
+            )
+            _config_logger.debug(msg, env_path, sorted_keys)
+
+        config.update(env_config)
+
+        if not config:
+            raise MissingConfigException(
+                f"No files of YAML or JSON format found in {base_path} or {env_path} matching"
+                f" the glob pattern(s): {[*self.config_patterns[key]]}"
+            )
+        return config
+
+    def __repr__(self):  # pragma: no cover
+        return (
+            f"OmegaConfLoader(conf_source={self.conf_source}, env={self.env}, "
+            f"config_patterns={self.config_patterns})"
+        )
+
+    def load_and_merge_dir_config(self, conf_path: str, patterns: Iterable[str]):
+        """Recursively load and merge all configuration files in a directory using OmegaConf,
+        which satisfy a given list of glob patterns from a specific path.
+
+        Args:
+            conf_path: Path to configuration directory.
+            patterns: List of glob patterns to match the filenames against.
+
+        Raises:
+            MissingConfigException: If configuration path doesn't exist or isn't valid.
+            ValueError: If two or more configuration files contain the same key(s).
+            ParserError: If config file contains invalid YAML or JSON syntax.
+
+        Returns:
+            Resulting configuration dictionary.
+
+        """
+        if not Path(conf_path).is_dir():
+            raise MissingConfigException(
+                f"Given configuration path either does not exist "
+                f"or is not a valid directory: {conf_path}"
+            )
+
+        paths = [
+            Path(each).resolve()
+            for pattern in patterns
+            for each in iglob(f"{str(conf_path)}/{pattern}", recursive=True)
+        ]
+        deduplicated_paths = set(paths)
+        config_files_filtered = [
+            path for path in deduplicated_paths if self._is_valid_config_path(path)
+        ]
+
+        config = {}
+        aggregate_config = []
+        seen_file_to_keys = {}  # type: Dict[Path, AbstractSet[str]]
+
+        for config_filepath in config_files_filtered:
+            try:
+                single_config = OmegaConf.load(config_filepath)
+                config = single_config
+            except (ParserError, ScannerError) as exc:
+                assert exc.problem_mark is not None
```

Comment:
What is this supposed to do? Could we get rid of the assert here?

Reply:
As far as I understand this is to verify whether we can get `exc.problem_mark.line` and `exc.problem_mark.column` for nicely formatting the error message. Perhaps the assert isn't necessary here.

Reply:
Ahh I think this might have been put there for the `mypy` linter:

```
kedro/config/omegaconf_config.py:224: error: Item "None" of "Optional[Mark]" has no attribute "line"
kedro/config/omegaconf_config.py:225: error: Item "None" of "Optional[Mark]" has no attribute "column"
Found 2 errors in 1 file (checked 265 source files)
```

I'll just add an ignore.